File: java/problem_solving/src/leetcode/MergeTwoSortedLists_21.java
package leetcode;

// #21 - Merge Two Sorted Lists

public class MergeTwoSortedLists_21 {

	static class ListNode {
		int val;
		ListNode next;

		ListNode() {
		}

		ListNode(int val) {
			this.val = val;
		}

		ListNode(int val, ListNode next) {
			this.val = val;
			this.next = next;
		}
	}

	public static void main(String[] args) {
		ListNode l3 = new ListNode(4);
		ListNode l2 = new ListNode(2, l3);
		ListNode l1 = new ListNode(1, l2);

		ListNode l6 = new ListNode(4);
		ListNode l5 = new ListNode(3, l6);
		ListNode l4 = new ListNode(1, l5);

		ListNode tmp = mergeTwoLists(l1, l4);
		while (tmp != null){
			System.out.println("tmp = " + tmp.val);
			tmp = tmp.next;
		}

	}

	public static ListNode mergeTwoLists(ListNode list1, ListNode list2) {
		ListNode tmp = new ListNode();
		ListNode head = tmp;
		// 하나라도 끝나면 종료
		while (list1 != null && list2 != null) {
			if (list1.val < list2.val) {
				tmp.next = list1;
				list1 = list1.next;
			} else {
				tmp.next = list2;
				list2 = list2.next;
			}

			tmp = tmp.next;
		}

		// 남은 부분 이어붙이기
		tmp.next = list1 != null? list1 : list2;
		return head.next;
	}
}

File: ruby/problem_solving/leetcode/MergeTwoSortedLists_21.rb
# 21 Merge Two Sorted Lists


class ListNode
  attr_accessor :val, :next

  def initialize(val = 0, _next = nil)
    @val = val
    @next = _next
  end
end

# @param {ListNode} list1
# @param {ListNode} list2
# @return {ListNode}
def merge_two_lists(list1, list2)
  return list2 if list1.nil?
  return list1 if list2.nil?

  if list1.val < list2.val
    list1.next = merge_two_lists(list1.next, list2)
    return list1
  else
    list2.next = merge_two_lists(list1, list2.next)
    return list2
  end

  # result = ListNode.new(-1)
  # tmp = result
  # while !list1.nil? && !list2.nil?
  #   l1_val = list1.val
  #   l2_val = list2.val
  #
  #   if l1_val < l2_val
  #     tmp.next = ListNode.new(l1_val)
  #     list1 = list1.next
  #   else
  #     tmp.next = ListNode.new(l2_val)
  #     list2 = list2.next
  #   end
  #
  #   tmp = tmp.next
  # end
  #
  # unless list1.nil?
  #   tmp.next = list1
  # else
  #   tmp.next = list2
  # end
  #
  # result.next
end

l2_3 = ListNode.new(4)
l2_2 = ListNode.new(3, l2_3)
list2 = ListNode.new(1, l2_2)
l1_3 = ListNode.new(4)
l1_2 = ListNode.new(2, l1_3)
list1 = ListNode.new(1, l1_2)

r = merge_two_lists(list1, list2)
while !r.nil?
  p r.val
  r = r.next
end